Source/FieldSolver/FiniteDifferenceSolver/EvolveB.cpp

```diff
@@ -0,0 +1,209 @@
+/* Copyright 2020 Remi Lehe
+ *
+ * This file is part of WarpX.
+ *
+ * License: BSD-3-Clause-LBNL
+ */
+
+#include "WarpXAlgorithmSelection.H"
+#include "FiniteDifferenceSolver.H"
+#ifdef WARPX_DIM_RZ
+#   include "FiniteDifferenceAlgorithms/CylindricalYeeAlgorithm.H"
+#else
+#   include "FiniteDifferenceAlgorithms/YeeAlgorithm.H"
+#   include "FiniteDifferenceAlgorithms/CKCAlgorithm.H"
+#   include "FiniteDifferenceAlgorithms/NodalAlgorithm.H"
+#endif
+#include <AMReX_Gpu.H>
+
+using namespace amrex;
+
+/**
+ * \brief Update the B field, over one timestep
+ */
+void FiniteDifferenceSolver::EvolveB (
+    std::array< std::unique_ptr<amrex::MultiFab>, 3 >& Bfield,
+    std::array< std::unique_ptr<amrex::MultiFab>, 3 > const& Efield,
+    amrex::Real const dt ) {
+
+   // Select algorithm (The choice of algorithm is a runtime option,
+   // but we compile code for each algorithm, using templates)
+#ifdef WARPX_DIM_RZ
+    if (m_fdtd_algo == MaxwellSolverAlgo::Yee){
+
+        EvolveBCylindrical <CylindricalYeeAlgorithm> ( Bfield, Efield, dt );
+
+#else
+    if (m_do_nodal) {
+
+        EvolveBCartesian <NodalAlgorithm> ( Bfield, Efield, dt );
+
+    } else if (m_fdtd_algo == MaxwellSolverAlgo::Yee) {
+
+        EvolveBCartesian <YeeAlgorithm> ( Bfield, Efield, dt );
+
+    } else if (m_fdtd_algo == MaxwellSolverAlgo::CKC) {
+
+        EvolveBCartesian <CKCAlgorithm> ( Bfield, Efield, dt );
+
+#endif
+    } else {
+        amrex::Abort("Unknown algorithm");
+    }
+
+}
+
+
+#ifndef WARPX_DIM_RZ
+
+template<typename T_Algo>
+void FiniteDifferenceSolver::EvolveBCartesian (
+    std::array< std::unique_ptr<amrex::MultiFab>, 3 >& Bfield,
+    std::array< std::unique_ptr<amrex::MultiFab>, 3 > const& Efield,
```

Comment:
Just to mention: I tried to make this 
```c++
std::array< std::unique_ptr<amrex::MultiFab const>, 3 > const& Efield
```
(because it's actually all we need) But it does not compile. Passing `int a;` in `f(const int)` works fine, but
```c++
std::array< std::unique_ptr<amrex::MultiFab const>, 3 > const
```
is a different type than 
```c++
std::array< std::unique_ptr<amrex::MultiFab>, 3 > const
```
That wasn't obvious to me.